### ISSUE TITLE
orafce update failed with ERROR: type "oracle.date" does not exist (orafce update from 3.7 to 4.9)

### ISSUE THREAD

**Issue by Post-User2024** [orafce/orafce#285]
Hi Team,

We have two database which are having the oafce extension current version of those extensions 3.7 before upgrading the postgresql V11 to Postgresql 15 we updated the orafce with make && make install after that update the extension like below.

 alter extension orafce update; - which is completed successfully on one database other database getting below error.

alter extension orafce update; --- failed to update on PostgreSQL V11
ERROR: type "oracle.date" does not exist

Could you please advise what is the cause of update issue?

**Comment by Post-User2024** [comment]
Since orafce was not update properly on one database!
Need your inputs can we go ahead with upgrade From PostgreSQL V11 to PostgreSQL V15? 
Is it a good idea to update orafce post upgrade of  PostgreSQL V15?

**Comment by Post-User2024** [comment]
Hello Team,

Could you please suggest!

**Comment by Post-User2024** [comment]
If your suspecting the problem with orafce version could you please share the latest development of the orafce?

**Comment by Post-User2024** [comment]
Before that can you please suggest why on one database extension update was success and other database failed to update the extension?

**Comment by okbob** [comment]
unfortunately I am not able repeat it. Usually is better to separate update of orafce from upgrade of Postgres. Orafce 4.2.2 should to support pg 11 and pg 15. So you can try to upgrade to 4.2.2 first. 

From reported error message is not possible to say, what is wrong. `oracle.date` type should be defined in 3.7 already.

**Comment by Post-User2024** [comment]
Hi Team,

We updated oracle orafce as below before the PostgreSQL upgrade, Post that upgrade was completed successfully.
Oracle fdw 1.2 
orafce - 4.13

On the PostgreSQL 15 we have updated oracle fdw and orafce extension with same level extensions aswell.

Just to verify the oracle fdw and orafce ran below select queries there were no output. That means extension having any issues?

Please share your inputs.

select orafce.rownum from my_foreign_table;
ERROR: relation "my_foreign_table" does not exist
LINE 1: select orafce.rownum from my_foreign_table;

select orafce.to_char(current_date,'YYYY-MM-DD');
ERROR: schema "orafce" does not exist
LINE 1: select orafce.to_char(current_date,'YYYY-MM-DD');

select * from my_foreign_tables limit 10;
ERROR: relation "my_foreign_tables" does not exist
LINE 1: select * from my_foreign_tables limit 10;

**Comment by Post-User2024** [comment]
select oracle_diag();
oracle_diag
oracle_fdw 2.8.0devel, PostgreSQL 15.8, Oracle client 18.3.0.0.0

**Comment by Post-User2024** [comment]
Hi 
Could you please share your inputs.

**Comment by Post-User2024** [comment]
Can you please advise how to validate extension is working or not as expected after the upgrade?

**Comment by Post-User2024** [comment]
any update on my request?

**Comment by okbob** [comment]
`Orafce` is just of set of functions and views created mostly in schema `oracle`.

```
(2024-10-22 20:54:57) postgres=# select * from pg_extension ;
┌───────┬─────────┬──────────┬──────────────┬────────────────┬────────────┬───────────┬──────────────┐
│  oid  │ extname │ extowner │ extnamespace │ extrelocatable │ extversion │ extconfig │ extcondition │
╞═══════╪═════════╪══════════╪══════════════╪════════════════╪════════════╪═══════════╪══════════════╡
│ 14640 │ plpgsql │       10 │           11 │ f              │ 1.0        │ ∅         │ ∅            │
│ 16385 │ orafce  │    16384 │         2200 │ f              │ 4.13       │ {16726}   │ {""}         │
└───────┴─────────┴──────────┴──────────────┴────────────────┴────────────┴───────────┴──────────────┘
(2 rows)

(2024-10-22 20:55:03) postgres=# select * from oracle.dual;
┌───────┐
│ dummy │
╞═══════╡
│ X     │
└───────┘
(1 row)

(2024-10-22 20:55:22) postgres=# select oracle.last_day('2024-10-22');
┌─────────────────────┐
│      last_day       │
╞═════════════════════╡
│ 2024-10-31 00:00:00 │
└─────────────────────┘
(1 row)
```

if you don't see this content, then some is wrong

**Comment by okbob** [comment]
The domain `oracle.date`  was introduced in `orafce 3.6`, but update script `orafce--3.5--3.6 looks weird. Maybe if your one database was initialized before `orafce 3.6`, then probably the functionality related to `oracle.date` is broken. Maybe you can try to execute script 

```
CREATE DOMAIN oracle.date AS timestamp(0);

CREATE OR REPLACE FUNCTION oracle.add_days_to_timestamp(oracle.date,integer)
RETURNS timestamp AS $$
SELECT $1 + interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.subtract (oracle.date, integer)
RETURNS timestamp AS $$
SELECT $1 - interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.add_days_to_timestamp(oracle.date,bigint)
RETURNS timestamp AS $$
SELECT $1 + interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.subtract (oracle.date, bigint)
RETURNS timestamp AS $$
SELECT $1 - interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.add_days_to_timestamp(oracle.date,smallint)
RETURNS timestamp AS $$
SELECT $1 + interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.subtract (oracle.date, smallint)
RETURNS timestamp AS $$
SELECT $1 - interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.add_days_to_timestamp(oracle.date,numeric)
RETURNS timestamp AS $$
SELECT $1 + interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.subtract (oracle.date, numeric)
RETURNS timestamp AS $$
SELECT $1 - interval '1 day' * $2;
$$ LANGUAGE SQL IMMUTABLE;

CREATE OR REPLACE FUNCTION oracle.subtract(oracle.date,oracle.date)
RETURNS double precision AS $$
SELECT date_part('epoch', ($1::timestamp - $2::timestamp)/3600/24);
$$ LANGUAGE SQL IMMUTABLE;

CREATE OPERATOR oracle.+ (
  LEFTARG   = oracle.date,
  RIGHTARG  = INTEGER,
  PROCEDURE = oracle.add_days_to_timestamp
);

CREATE OPERATOR oracle.- (
  LEFTARG   = oracle.date,
  RIGHTARG  = INTEGER,
  PROCEDURE = oracle.subtract
);

CREATE OPERATOR oracle.+ (
  LEFTARG   = oracle.date,
  RIGHTARG  = bigint,
  PROCEDURE = oracle.add_days_to_timestamp
);

CREATE OPERATOR oracle.- (
  LEFTARG   = oracle.date,
  RIGHTARG  = bigint,
  PROCEDURE = oracle.subtract
);

CREATE OPERATOR oracle.+ (
  LEFTARG   = oracle.date,
  RIGHTARG  = smallint,
  PROCEDURE = oracle.add_days_to_timestamp
);

CREATE OPERATOR oracle.- (
  LEFTARG   = oracle.date,
  RIGHTARG  = smallint,
  PROCEDURE = oracle.subtract
);

CREATE OPERATOR oracle.+ (
  LEFTARG   = oracle.date,
  RIGHTARG  = numeric,
  PROCEDURE = oracle.add_days_to_timestamp
);

CREATE OPERATOR oracle.- (
  LEFTARG   = oracle.date,
  RIGHTARG  = numeric,
  PROCEDURE = oracle.subtract
);

CREATE OPERATOR oracle.- (
  LEFTARG   = oracle.date,
  RIGHTARG  = oracle.date,
  PROCEDURE = oracle.subtract
);
```
before you will try to upgrade to some higher.  You can work just on system where domain `oracle.date` was not created yet

```
(2024-10-22 21:34:21) postgres=# \dT oracle.date
          List of data types
┌────────┬─────────────┬─────────────┐
│ Schema │    Name     │ Description │
╞════════╪═════════════╪═════════════╡
│ oracle │ oracle.date │ ∅           │
└────────┴─────────────┴─────────────┘
(1 row)
```
not like this - `\dt oracle.date` should to return zero rows

**Comment by Post-User2024** [comment]
Thanks for the quick response.

This script we need to execute that database where we had an issue before the upgrade plan?  please confirm.

**Comment by okbob** [comment]
This script should be executed before upgrade

Dne út 22. 10. 2024 21:43 uživatel Post-User2024 ***@***.***>
napsal:

**Comment by Post-User2024** [comment]
lets say have 10 databases under the PG cluster.

4 dbs are using the orafce extension we have to run the above script  on 4 dbs (or) entire cluster like connecting each database run that script? 

Post executing that script can we update the orafce extension from 3.7 to 4.13 before upgrading v11 to v15?

Please share your view on this.

**Comment by Post-User2024** [comment]
Getting below error when i tried to restore the v11 database dump on v15.

Please review and advise. (orafce extension version is 4.13 on v15).

nohup: ignoring input
pg_restore: error: could not execute query: ERROR: extension "pg_hint_plan" is not available
DETAIL: Could not open extension control file "/usr/pgsql-15/share/extension/pg_hint_plan.control": No such file or directory.
HINT: The extension must first be installed on the system where PostgreSQL is running.
Command was: CREATE EXTENSION IF NOT EXISTS pg_hint_plan WITH SCHEMA hint_plan;

pg_restore: error: could not execute query: ERROR: function to_char(integer) does not exist

**Comment by okbob** [comment]
st 23. 10. 2024 v 7:44 odesílatel Post-User2024 ***@***.***>
napsal:

> lets say have 10 databases under the PG cluster.
>
> 4 dbs are using the orafce extension we have to run the above script on 4
> dbs (or) entire cluster like connecting each database run that script?
>
this is per database


> Post executing that script can we update the orafce extension from 3.7 to
> 4.13 before upgrading v11 to v15?
>

you should to execute this script before any upgrade from 3.7

**Comment by okbob** [comment]
st 23. 10. 2024 v 7:45 odesílatel Post-User2024 ***@***.***>
napsal:

> Getting below error when i tried to restore the v11 database dump on v15.
>
> Please review and advise. (orafce extension version is 4.13 on v15).
>
> nohup: ignoring input
> pg_restore: error: could not execute query: ERROR: extension
> "pg_hint_plan" is not available
> DETAIL: Could not open extension control file
> "/usr/pgsql-15/share/extension/pg_hint_plan.control": No such file or
> directory.
> HINT: The extension must first be installed on the system where PostgreSQL
> is running.
> Command was: CREATE EXTENSION IF NOT EXISTS pg_hint_plan WITH SCHEMA
> hint_plan;
>
> pg_restore: error: could not execute query: ERROR: function
> to_char(integer) does not exist
>
it is common problem - the best is remove any extension that can be removed
before Postgres upgrade - mainly if you do jump from unsupported release

if it is possible, update other extensions to same version on old and new
instances before Postgres upgrade. For orafce you cannot to go to the last
version immediately, because this version is not available for Postgres 11.

**Comment by Post-User2024** [comment]
Lets take one scenario

My upgrade was failed (From v11 to V15) due to orafce 4.2.2 extension which is samelevel on both V11 and v15.

In this case can we drop the orafce extension then proceed for the upgrade again post upgrade what are all the changes we have to face during the orafce extension creation with 4.2.2?


Is there any way to get the orafce related functions deatilas with function complete syntax of those function's before upgrade if something went wrong can we re-create those functions on the V15 manually?

If it is possible please help with commands and queries.

**Comment by okbob** [comment]
st 23. 10. 2024 v 8:39 odesílatel Post-User2024 ***@***.***>
napsal:

> Lets take one scenario
>
> My upgrade was failed (From v11 to V15) due to orafce 4.2.2 extension
> which is samelevel on both V11 and v15.
>
> In this case can we drop the orafce extension then proceed for the upgrade
> again post upgrade what are all the changes we have to face during the
> orafce extension creation with 4.2.2?
>
> Is there any way to get the orafce related functions deatilas with
> function complete syntax of those function's before upgrade if something
> went wrong can we re-create those functions on the V15 manually?
>
> If it is possible please help with commands and queries.
>
if it is possible to drop orafce, do it. And install fresh orafce on v15

**Comment by Post-User2024** [comment]
Have executed the dT oracle.date command on production there is no output looks like orafce was broken for this database and  as you said we have to run that script before upgrade the orafce  from 3.7 to 4.2.4 before proceeding the actual postgres upgrade From v11 To v15.

Could you please inputs on this.

Once the v11 to v15 upgrade completed with any issues then we can plan for the orafce upgrade From 4.2.2 To 4.13/latest version.

 \dT oracle.date
     List of data types
 Schema | Name | Description
--------+------+-------------
(0 rows)

select * from pg_extension;
      extname       | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition
--------------------+----------+--------------+----------------+------------+-----------+--------------
 orafce             |    16387 |         2200 | f              | 3.7        |           |

**Comment by okbob** [comment]
st 23. 10. 2024 v 9:37 odesílatel Post-User2024 ***@***.***>
napsal:

> Have executed the dT oracle.date command on production there is no output
> looks like orafce was broken for this database and as you said we have to
> run that script before upgrade the orafce from 3.7 to 4.2.4 before
> proceeding the actual postgres upgrade From v11 To v15.
>

you can try it on some clone. and yes, you should to execute the script
before upgrade from 3.7. When there is not definition of oracle.date domain
then upgrade to higher version should to fail.

**Comment by Post-User2024** [comment]
Just for testing we started restoring the v11 database dump on v15 getting below error while restore.

Source v11 - database orafce -3.7 version.
Target v15 - database orafce - 4.14 version.

pg_restore: error: could not execute query: ERROR:  type "public.varchar2" does not exist.

How to fix this error?

**Comment by Post-User2024** [comment]
Can we try with create extension orafce schema public will that works?

**Comment by okbob** [comment]
st 23. 10. 2024 v 19:07 odesílatel Post-User2024 ***@***.***>
napsal:

> Just for testing we started restoring the v11 database dump on v15 getting
> below error while restore.
>
> Source v11 - database orafce -3.7 version.
> Target v15 - database orafce - 4.14 version.
>
> pg_restore: error: could not execute query: ERROR: type "public.varchar2"
> does not exist.
>
> How to fix this error?
>

It is hard to do it. In old versions of Orafce, this type was in public
schema. But in 3.22 was migrated to oracle schema.

Just cannot to work dump with one version and restore with other.

**Comment by okbob** [comment]
st 23. 10. 2024 v 19:10 odesílatel Post-User2024 ***@***.***>
napsal:

> Can we try with create extension orafce schema public will that works?
>

no

**Comment by Post-User2024** [comment]
Was there any workaround this kind of issues?

Because if we downgrade any of the oracle fdw and orafce extension PostgreSQL upgrade dry run creating the varchar and create foreign table errors (From v11 To v15).

**Comment by okbob** [comment]
st 23. 10. 2024 v 19:32 odesílatel Post-User2024 ***@***.***>
napsal:

> Was there any workaround this kind of issues?
>
you can try to manually edit dump and replace public.varchar2 to
oracle.varchar2

**Comment by Post-User2024** [comment]
I saw other issues while restoring on the same(From v11 To v15).

pg_restore: error: could not execute query: ERROR:  function to_char(bigint) does not exist

Can you please share any around for this error?

I'm just thinking for next restore test will take dump format in .sql file then will refer the restore current error log file where we want change the code conversion will change accordingly then proceed for the next restore test?

This way can do ? Please share your inputs.

**Comment by okbob** [comment]
čt 24. 10. 2024 v 7:25 odesílatel Post-User2024 ***@***.***>
napsal:

> I saw other issues while restoring on the same(From v11 To v15).
>
> pg_restore: error: could not execute query: ERROR: function
> to_char(bigint) does not exist
>
it can be related to orafce. there are lot of functions that was moved from
public to oracle schema.



> Can you please share any around for this error?
>
> I'm just thinking for next restore test will take dump format in .sql file
> then will refer the restore current error log file where we want change the
> code conversion will change accordingly then proceed for the next restore
> test?
>

yes - it was my previous advice

**Comment by Post-User2024** [comment]
Can you please advise what exactly we want to change for pg_restore: error: could not execute query: ERROR: function to_char(bigint) does not exist issue?

There is no public (or) oracle looks like we want to change the data type?

**Comment by Post-User2024** [comment]
- move objects from pg_catalog and from public schema to schema oracle 
- 
- orafce 3.21--3.22.sql
EXECUTE $_$ALTER FUNCTION pg_catalog.to_char(bigint) SET SCHEMA oracle$_$;

orafce 4.14.sql

CREATE FUNCTION oracle.to_char(num bigint)
RETURNS text
AS 'MODULE_PATHNAME','orafce_to_char_int8'
LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
COMMENT ON FUNCTION oracle.to_char(bigint) IS 'Convert number to string';

**Comment by okbob** [comment]
Unfortunately, this issue is related to long time unsupported releases of PostgreSQL, and it is very difficult to fix it.

**Comment by okbob** [comment]
I was wrong - oracle.date was introduced in orafce 3.0.7. The update scripts are supported from orafce 3.2+.

probably your system was initially installed before release 3.1 and probably by execution of registration sql script (the method used before introduction of CREATE EXTENSION command). I found a error in orafce--unpackaged--3.x.sql - there is missing to assign domain oracle.date to extension orafce. So probably when somebody did migration from old method to new method, and upgrade by pg_dump, then the domain oracle.date was removed from your database. Restore from pg_dump is fault tolerant, but the ALTER EXTENSION UPDATE not. Unfortunately, the error was done ten years ago, and now it is not possible to fix it on orafce side. Buggy files are removed, and old method (of installation) is not supported.
